expand.go

```diff
@@ -0,0 +1,217 @@
+// a literal '$'.
+func (e *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int, error) {
+	var i int
+	var j int
+
+	for i < len(src) {
+
+		if j == len(dst) {
+			return j, i, transform.ErrShortDst
+		}
+
+		end := bytes.IndexByte(src[i:], '$')
+
+		if end == -1 {
+			// src does not contain '$', copy into dst
+			cnt := copy(dst[j:], src[i:])
+			i += cnt
+			j += cnt
+			continue
+		} else if end > 0 {
+			// copy chars preceding '$' from src to dst
+			cnt := copy(dst[j:], src[i:i+end])
+			i += cnt
+			j += cnt
+
+			if j == len(dst) {
+				return j, i, transform.ErrShortDst
+			}
+
+			// fallthrough
+		}
+
+		// src[i] now points to '$', j < len(dst)
+
+		// If we're at the end of src, but we found a starting
+		// token, return ErrShortSrc, unless we're also at EOF,
+		// in which case just copy it dst.
+		if i+1 == len(src) {
+			if atEOF {
+				dst[j] = src[i]
+				i++
+				j++
+				continue
+			}
+			return j, i, transform.ErrShortSrc
+		}
+
+		// At this point we know that src[i+1] is populated.
+
+		// If this token sequence represents the special '$$'
+		// sequence, emit a '$' into dst.
+		if src[i+1] == '$' {
+			dst[j] = src[i]
+			i += 2
+			j++
+			continue
+		}
+
+		var token []byte
+
+		// Start of bracketed token ${foo}
+		if src[i+1] == '{' {
+			end = bytes.IndexByte(src[i+2:], '}')
+			if end == -1 {
+				if atEOF {
+					// No closing bracket and we're at
+					// EOF, so it's not a valid bracket
+					// expression.
+					cnt := copy(dst[j:], src[i:])
+					if i+cnt == len(src) {
+						i += cnt
+						j += cnt
+						continue
+					}
+
+					// If dst didn't have enough space
+					// return ErrShortDst
+					return j, i, transform.ErrShortDst
+				}
+
+				// Otherwise, we need more bytes in src
+				return j, i, transform.ErrShortSrc
+			}
+
+			// Adjust end so it points to the byte
+			// immediately after the closing '}'
+			end += i + 3
+
+			token = src[i+2 : end-1]
+		} else { // Else start of non-bracketed token $foo
+			if !isShellNameFirstChar(src[i+1]) {
+				// If it doesn't conform to the naming
+				// rules for shell variables, do not
+				// try to expand, just copy to dst.
+				dst[j] = src[i]
+				i++
+				j++
+				continue
+			}
+
+			end = bytesIndexCFunc(src[i+2:], isShellNameChar)
+
+			if end == -1 {
+				// Reached the end of src without finding
+				// end of shell variable
+				if !atEOF {
+					// We need more bytes in src
+					return j, i, transform.ErrShortSrc
+				}
+				end = len(src)
+			} else {
+				// Adjust end so it points to the byte
+				// immediately after the token
+				end += i + 2
+			}
+
+			token = src[i+1 : end]
+		}
+
+		replacement, err := e.expand(string(token))
+		if err != nil {
+			return j, i, err
+		}
+
+		cnt := copy(dst[j:], replacement)
+		if cnt == len(replacement) {
+			i = end
+			j += cnt
+			continue
+		}
+
+		// If we couldn't copy the replacement string into dst,
+		// we need more room in dst.
+
+		return j, i, transform.ErrShortDst
```

Comment:
Can we test for this by providing a buffer that is too short for a very large replacement string?

Reply:
There should already be a test that triggers this code path.  Ultimately the text/transform package currently uses a buffer that is 4096 bytes, so a substitution may not be longer than that.